rust/otap-dataflow/crates/otap/src/otap_batch_processor.rs

```diff
@@ -0,0 +1,698 @@
+    ///
+    /// MVP: threshold is based on count (number of messages). In future, switch to real
+    /// row counts for OtapArrowRecords.
+    fn size_triggers_emission(&self) -> bool {
+        self.current.len() >= self.config.send_batch_size
+    }
+
+    /// Partition buffered items by signal type when possible and emit them downstream.
+    /// Items that cannot be converted to OtapArrowRecords are passed through as-is.
+    async fn flush_current(
+        &mut self,
+        effect: &mut local::EffectHandler<OtapPdata>,
+    ) -> Result<(), EngineError<OtapPdata>> {
+        if self.current.is_empty() {
+            return Ok(());
+        }
+        let mut drained: Vec<OtapPdata> = std::mem::take(&mut self.current);
+
+        // Separate items that can be converted to OtapArrowRecords from passthrough
+        let mut convertible: Vec<OtapArrowRecords> = Vec::new();
+        let mut passthrough: Vec<OtapPdata> = Vec::new();
+        for item in drained.drain(..) {
+            match OtapArrowRecords::try_from(item.clone()) {
+                Ok(rec) => convertible.push(rec),
+                Err(_) => passthrough.push(item),
+            }
+        }
+
+        // If nothing convertible, just emit passthrough items and return.
+        if convertible.is_empty() {
+            for item in passthrough {
+                effect.send_message(item).await?;
+            }
+            return Ok(());
+        }
+
+        // Build output batches using low-level batching (split + concatenate per type)
+        let max = NonZeroU64::new(self.config.send_batch_max_size as u64);
+        let output_batches = match low_make_output_batches(max, convertible.clone()) {
+            Ok(v) => v,
+            Err(e) => {
+                // Fall back to passthrough of convertible records on error
+                effect
+                    .info(&format!(
+                        "OTAP batch processor: low-level batching failed: {e}; falling back"
+                    ))
+                    .await;
+                convertible
+            }
+        };
+
+        // Emit converted (records) first, then passthrough
+        for records in output_batches {
+            let pdata: OtapPdata = records.into();
+            effect.send_message(pdata).await?;
+        }
+        for item in passthrough {
+            effect.send_message(item).await?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait(?Send)]
+impl local::Processor<OtapPdata> for OtapBatchProcessor {
+    async fn process(
+        &mut self,
+        msg: Message<OtapPdata>,
+        effect: &mut local::EffectHandler<OtapPdata>,
+    ) -> Result<(), EngineError<OtapPdata>> {
+        match msg {
+            Message::Control(ctrl) => {
+                match ctrl {
+                    otap_df_engine::control::NodeControlMsg::TimerTick { .. } => {
+                        // Flush any buffered items on timer
+                        self.flush_current(effect).await
```

Comment:
In the OTLP batch processor, we don't always flush on the timer tick. It looks like we flush if some threshold has been reached since the last time we saw an incoming batch
https://github.com/open-telemetry/otel-arrow/blob/8c4473eefcd9f1107e49a56bb26d3b74d4c97ab0/rust/otap-dataflow/crates/otlp/src/otlp_batch_processor.rs#L441-L463

Should we do something similar here?

Reply:
i have updated to follow these semantics